docs/decisions/0017-generalized-containers.rst

```diff
@@ -0,0 +1,70 @@
+===========================================
+
+- A container is designed as a generalized capability to hold different types of content.
+- A container is a publishable content type that holds other content types through a parent-child relationship.
+- A container application will offer shared definitions for use by other container types.
+
+2. Container Types and Content Constraints
+==========================================
+
+- A container marks any PublishableEntity, such as sections, subsections, units, or any other custom content type, as a type that can hold other content.
+- Containers can be nested within other containers, allowing for complex content structures.
+- Containers might be of different types, with each type potentially having different restrictions on the type of content it can hold but that will not be enforced by containers.
+- Content restrictions for containers are implemented at the app layer, allowing specific container types, like Unit, to limit their members to particular content types (e.g., only Components).
+- The course hierarchy Course > Section > Subsection > Unit will be implemented as relationships between containers, with each level acting as a container that holds other content. The hierarchy will be enforced by the content restrictions of each particular container but allowed to be overridden to support `0002-content-flexibility.rst`_.
+- Containers will follow extensibility principles in `0003-content-extensibility.rst`_ for creating new container types or subtypes.
+
+3. Container Members and Relationships
+=======================================
+
+- The members of a container can be any type of publishable content.
+- Members within a container are maintained in a specific order as an ordered list.
+- Containers represent their content hierarchy through a structure that defines parent-child relationships between the container and its members.
+- The structure defining these relationships is anonymous, so it can only be referenced through the container.
+- Containers can hold both static and dynamically generated content, including user-specific variations.
+- Each container holds different states of its members (user-defined, initial, and frozen final state) to support rollback operations.
+- Members can be added or removed from a container, and the container will maintain the state of the content for the previous version (frozen final state).
+- The initial state of a container is immutable.
+- When a container's structure changes, e.g., when a new member is added, the user-defined state of the container is updated with the new members list.
+- Containers support both fixed and version-agnostic references for members, allowing members to be pinned to a specific version or set to reference the latest draft or published state.
+- The latest draft or published states can be referenced by setting the version to ``None``, avoiding the need for new instances on each update.
```

Comment:
Is there a difference between specifying "latest draft" and specifying "latest published" or is there just a way to say "latest" (version=None), and whether that is draft or published depends on _______ ?


EDIT: OK, from looking at the code it appears that each container (via `EntityListRow`) can specify _both_ a `draft_version` and a `published_version` at the same time, separately. And each can be either pinned or unpinned. I guess I'm very unclear on how this works when the `ContainerEntityVersion` itself is versioned and has a draft + published version.

 If I have a draft `ContainerEntityVersion`, what do its `EntityListRow`s' `published_version`s mean? And when I publish it, so it's now a published `ContainerEntityVersion`, what do its `EntityListRow`s' `draft_version`s mean?

Reply:
> If I have a draft `ContainerEntityVersion`, what do its `EntityListRow`s' `published_version`s mean? And when I publish it, so it's now a published `ContainerEntityVersion`, what do its `EntityListRow`s' `draft_version`s mean?

I originally had it with just `version`, but I vaguely remember thinking that I had to add `draft_version` separately from `published_version` to address some edge case... which I have totally forgotten. And maybe was wrong-headed to begin with. 😛 Let's take it back out for now–if said weird edge case was real, I'm sure we'll run into it in the not-too-distant future, and deal with it then.

Reply:
@ormsbee: during our latest discussion, you mentioned that having both, mainly the draft version, covered the CCX use cases where we don't necessarily control the latest draft version, so we'd want to pin it for the author's context. Do you think we should cover that case later down the line?

We also discussed that having both would help us know which published/draft versions were in the frozen list when creating the next version, but I think that the use case could be covered by having a single reference to the current version draft or published. Would you agree?

Reply:
> @ormsbee: during our latest discussion, you mentioned that having both, mainly the draft version, covered the CCX use cases where we don't necessarily control the latest draft version, so we'd want to pin it for the author's context. Do you think we should cover that case later on?

I think I was mistaken there, because if the author wanted to pin to a specific version, they could do so, and then when they decided to undo that pin to get something later, they could do so with a new version of the container–no need to keep the publish versioned separate.

> We also discussed that having both would help us know which published/draft versions were in the frozen list when creating the next version, but I think that the use case could be covered by having a single reference to the current version draft or published. Would you agree?

Yeah, I agree.

I think where my head might have been going is what happens in the following scenario:

1. UnitVersion UV1 has an unpinned reference to Component C1.
2. C1 has ComponentVersions CV1 and CV2. CV1 is the current published version, while CV2 is the draft version.
3. C1 is soft-deleted, forcing UV1 to pin down its references to the last version. But does it pin to CV1 or CV2?

-----

So I'm still kicking it around in my head, but I'm trying to see how things line up with Unit/Container modeling if we don't force new container versions to be created when things get deleted, and instead filter out the deleted stuff at read time. I think it could get rid of a lot of the bookkeeping needs for the model. I'm trying to sketch this out this evening to see if it fits together in a reasonable way...